cpp/algo_module/algorithm/algo.cpp

```diff
@@ -1,5 +1,132 @@
+    bool starts_with = rel_type.starts_with('<');
+    bool ends_with = rel_type.ends_with('>');
+
+    if (rel_type.size() == 1) {
+      if (starts_with) {
+        any_incoming_ = true;
+      } else if (ends_with) {
+        any_outgoing_ = true;
+      } else {
+        rel_direction_[rel_type] = RelDirection::kAny;
+      }
+      continue;
+    }
+
+    if (starts_with && ends_with) {  // <type>
+      rel_direction_[rel_type.substr(1, rel_type.size() - 2)] = RelDirection::kBoth;
+    } else if (starts_with) {  // <type
+      rel_direction_[rel_type.substr(1)] = RelDirection::kIncoming;
+    } else if (ends_with) {  // type>
+      rel_direction_[rel_type.substr(0, rel_type.size() - 1)] = RelDirection::kOutgoing;
+    } else {  // type
+      rel_direction_[rel_type] = RelDirection::kAny;
+    }
+  }
+}
+
+Algo::RelDirection Algo::PathFinder::GetDirection(const std::string &rel_type) const {
+  auto it = rel_direction_.find(rel_type);
+  if (it == rel_direction_.end()) {
+    return RelDirection::kNone;
+  }
+  return it->second;
+}
+
+void Algo::PathFinder::DFS(const mgp::Node &curr_node, mgp::Path &curr_path, std::unordered_set<int64_t> &visited) {
+  if (curr_node.Id() == end_node_id_) {
+    auto record = record_factory_.NewRecord();
+    record.Insert(std::string(kResultAllSimplePaths).c_str(), curr_path);
+    return;
+  }
+
+  if (static_cast<int64_t>(curr_path.Length()) == max_length_) {
+    return;
+  }
+
+  visited.insert(curr_node.Id().AsInt());
+  std::unordered_set<int64_t> seen;
+
+  auto iterate = [&visited, &seen, &curr_path, this](mgp::Relationships relationships, bool outgoing) {
+    RelDirection direction = (outgoing ? RelDirection::kOutgoing : RelDirection::kIncoming);
+
+    for (const auto relationship : relationships) {
+      auto next_node_id = outgoing ? relationship.To().Id().AsInt() : relationship.From().Id().AsInt();
+
+      if (visited.contains(next_node_id)) {
+        continue;
+      }
+
+      auto type = std::string(relationship.Type());
+      auto wanted_direction = GetDirection(type);
+
+      if (wanted_direction == RelDirection::kAny || wanted_direction == direction ||
+          (any_incoming_ && direction == RelDirection::kIncoming) ||
+          (any_outgoing_ && direction == RelDirection::kOutgoing)) {
+        curr_path.Expand(relationship);
+        DFS(outgoing ? relationship.To() : relationship.From(), curr_path, visited);
+        curr_path.Pop();
+      } else if (wanted_direction == RelDirection::kBoth) {
+        if (outgoing && seen.contains(relationship.To().Id().AsInt())) {
+          curr_path.Expand(relationship);
+          DFS(relationship.To(), curr_path, visited);
+          curr_path.Pop();
+        } else if (!outgoing) {
+          seen.insert(relationship.From().Id().AsInt());
+        }
+      }
+    }
+  };
+
+  iterate(curr_node.InRelationships(), false);
+  iterate(curr_node.OutRelationships(), true);
+  visited.erase(curr_node.Id().AsInt());
```

Comment:
Try to check here if user sent at all InRelationship, and if not that we don't expand on in relationships